CASSANDRA-17031: Add support for PEM formatted key material for the SSL context

maulin-vasavada opened the pull request.

Commit:
ensure hint window is persistent across restarts of a node

patch by Kurt Greaves; reviewed by Brandon Williams, Mick Semb Wever and Stefan Miklosovic for CASSANDRA-14309

Commit:
Add blocking pendingranges calculation when keyspace is added

Patch by Sumanth Pasupuleti, reviewed by Alex Petrov for CASSANDRA-15433

Commit:
Log more information when a node runs out of commitlog space

Patch by Boris Onufriyev; review by Ariel Weisberg, Benjamin Lerer and
Brandon Williams for CASSANDRA-11323

bdeggleston committed:
Actively update auth caches in the background

Patch by Blake Eggleston; reviewed by Sam Tunnicliffe, Jason Brown, and Caleb Rackliffe for CASSANDRA-16957

Co-authored-by: Blake Eggleston <bdeggleston@gmail.com>
Co-authored-by: Josh McKenzie <jmckenzie@apache.org>

jmckenzie-dev committed:
Add a Denylist to block reads and writes on specific partition keys

Patch by Josh McKenzie, reviewed by Aleksei Zotov and Sumanth Pasupuleti for CASSANDRA-12106

Co-authored by Josh McKenzie <jmckenzie@apache.org>
Co-authored by Sam Overton

azotcsit committed:
Make nodetool compactionstats and sstable_tasks consistent

patch by Aleksei Zotov; reviewed by Stefan Miklocovic and Brandon Williams for CASSANDRA-16976

blerer committed:
Allow to grant permission for all tables in a keyspace

Patch by Benjamin Lerer; Review by Andres de la Peña and Ekaterina Dimitrova for CASSANDRA-17027

In some cases it is useful to prevent users to alter or drop a keyspace
while allowing them to create new tables.
This patch add support for a new DataResource below KEYSPACE but above
TABLE. The syntax to grant permission at this level in ALL TABLES IN
KEYSPACE.

driftx committed:
Upgrade to snakeyaml 1.26

Patch by brandonwilliams; reviewed by bereng for CASSANDRA-17028

frankgh committed:
Allow GRANT/REVOKE multiple permissions in a single statement

patch by Francisco Guerrero; reviewed by Benjamin Lerer and Yifan Cai for CASSANDRA-17030

This commit allows GRANT/REVOKE statement to support multiple permissions with a single
statement. For example,

```
GRANT MODIFY, SELECT ON KEYSPACE field TO manager;
GRANT ALTER, DROP ON ROLE role1 TO role2;
```

Commit:
first cut

Commit:
adding PEMReaderTest

Commit:
Matching config names with the CEP and adding tests for unencrypted key

krummas committed:
Log time spent writing keys during compaction

Patch by Marcus Eriksson; reviewed by Sam Tunnicliffe for CASSANDRA-17037

Co-authored-by: Marcus Eriksson <marcuse@apache.org>
Co-authored-by: Josh McKenzie <jmckenzie@apache.org>

driftx committed:
Push initial connection messages to trace

Patch by brandonwilliams; reviewed by bereng for CASSANDRA-17038

Commit:
Adding documentation, tests for PEM and JKS combination and test for hot reloading

adelapena committed:
Save CircleCI resources with optional test jobs

patch by Andrés de la Peña; reviewed by Ekaterina Dimitrova and Berenguer Blasi for CASSANDRA-16882

adelapena committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

adelapena committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

adelapena committed:
Merge branch 'cassandra-4.0' into trunk

blerer committed:
Merge branch cassandra-3.0 into cassandra-3.11

blerer committed:
Merge branch cassandra-3.11 into cassandra-4.0

blerer committed:
Merge branch cassandra-4.0 into trunk

ifesdjeen committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

ifesdjeen committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

ifesdjeen committed:
Merge branch 'cassandra-4.0' into trunk

driftx committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

driftx committed:
Merge branch 'cassandra-4.0' into trunk

Commit:
Adding K8s PEM based SslContextFactory example

subkanthi committed:
Add unix time conversion functions

Patch by Kanthi Subramanian; reviewed by Benjamin Lerer and Brandon
Williams for CASSANDRA-17029

driftx committed:
Push connection reset by peer messages to trace

Patch by brandonwilliams; reviewed by bereng for CASSANDRA-16954

Commit:
Adding support for DSA and EC based private keys

adelapena committed:
Use dedicated executors for CircleCI repeated test jobs

patch by Andrés de la Peña; reviewed by Ekaterina Dimitrova for CASSANDRA-17043

adelapena committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

adelapena committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

adelapena committed:
Merge branch 'cassandra-4.0' into trunk

Commit:
Update PEMReader.java

bereng committed:
Prevent broken concurrent schema read/writes

patch by Berenguer Blasi; reviewed by Caleb Rackliffe for CASSANDRA-16996

Co-authored-by: Berenguer Blasi <berenguerblasi@gmail.com>
Co-authored-by: Caleb Rackliffe <calebrackliffe@gmail.com>

driftx committed:
Merge branch 'cassandra-4.0' into trunk

driftx committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

driftx committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

driftx committed:
Merge branch 'cassandra-4.0' into trunk

driftx committed:
Set fork mode to class in IntelliJ default test config

Patch by Jacek Lewandowski; reviewed by brandonwilliams and marcuse for
CASSANDRA-17053

driftx committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

driftx committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

driftx committed:
Merge branch 'cassandra-4.0' into trunk

subkanthi committed:
Remove duplicate toCQLString in ReadCommand

Patch by Kanthi Subramanian; reviewed by brandonwilliams and maedhroz
for CASSANDRA-17023

dcapwell committed:
v4+ protocol did not clean up client warnings, which caused leaking the state

patch by David Capwell; reviewed by Caleb Rackliffe, Jon Meredith, Sam Tunnicliffe for CASSANDRA-17054

bereng committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

bereng committed:
Merge branch 'cassandra-4.0' into trunk

Gerrrr committed:
Fix failure handling in inter-node communication

patch by Aleksandr Sorokoumov; reviewed by Andrés de la Peña and Paulo Motta for CASSANDRA-16334

adelapena committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

adelapena committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

adelapena committed:
Merge branch 'cassandra-4.0' into trunk

adelapena committed:
Allow users to view permissions of the roles they created

patch by Robert Stupp, Andrés de la Peña and Aleksei Zotov; reviewed by Aleksei Zotov and Benjamin Lerer for CASSANDRA-16902

Co-authored-by: Robert Stupp <snazy@snazy.de>
Co-authored-by: Andres de la Pena <a.penya.garcia@gmail.com>
Co-authored-by: Aleksei Zotov <azotcsit@gmail.com>

adelapena committed:
Merge branch 'cassandra-3.0' into cassandra-3.11

adelapena committed:
Merge branch 'cassandra-3.11' into cassandra-4.0

adelapena committed:
Merge branch 'cassandra-4.0' into trunk

bereng committed:
ViewComplexTest hardening

patch by Andres de la Peña, Berenguer Blasi; reviewed by Andres de la Peña for CASSANDRA-17070

Co-authored-by: Andres de la Peña <a.penya.garcia@gmail.com>
Co-authored-by: Berenguer Blasi <berenguerblasi@gmail.com>

jmckenzie-dev committed:
Ninja-fix: CHANGES.txt and NEWS.txt for denylisting (CASSANDRA-12106)

dcapwell committed:
commit log was switched from non-daemon to daemon threads, which causes the JVM to exit in some case as no non-daemon threads are active

patch by David Capwell, Sam Tunnicliffe; reviewed by Sam Tunnicliffe for CASSANDRA-17085

JeetKunDoug committed:
Backward compatibility for CQLSSTableWriter Date fields

Patch by Doug Rohrer; reviewed by Brandon Williams and Sam Tunnicliffe
for CASSANDRA-17117

dcapwell committed:
coordinator_read_size had wrong name in cassandra.yaml

patch by Ekaterina Dimitrova, reviewed by David Capwell for CASSANDRA-17118

bereng committed:
Merge branch 'cassandra-4.0' into trunk

belliottsmith committed:
Forbid other Future implementations with checkstyle

patch by Benedict Elliott Smith; reviewed by David Capwell for CASSANDRA-17055

beobal committed:
Merge branch 'cassandra-4.0' into trunk

smiklosovic commented on examples/ssl-factory/test/unit/org/apache/cassandra/security/KubernetesSecretsPEMSslContextFactoryTest.java:
```diff
@@ -0,0 +1,436 @@
+        }
+    }
+
+    private static final String private_key =
```
cant be this somehow fetched from a helper file in test resources?

maulin-vasavada replied:
Done

smiklosovic commented on examples/ssl-factory/test/unit/org/apache/cassandra/security/KubernetesSecretsPEMSslContextFactoryTest.java:
```diff
@@ -0,0 +1,436 @@
+            }
+            else
+            {
+                logger.info("Configuration has env variable {} with value {}. Will use that.",
```
env -> environment

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+
+        ConfigKey(String keyName)
+        {
+            this.keyName=keyName;
```
pay attention to the right formatting, I am not sure if your IDE is formatting it just that way, I see this is the re-occuring issue when I am reviewing your patches. Same for stuff like "for(String s : strings)", there needs to be space between "for" and opening bracket and similar.

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+        {
+            Set<String> valueSet = new HashSet<>();
+            ConfigKey[] values = values();
+            for(ConfigKey key: values) {
```
see?

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+        keyPassword = getString(ConfigKey.KEY_PASSWORD.getKeyName());
+        if (StringUtils.isEmpty(keyPassword))
+        {
+            keyPassword = keystore_password;
```
why is that "keystore_password"? I think we are just using camel cases here.

maulin-vasavada replied:
We can tie this comment to the following comment around the same and resolve this with that.

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+        else if (!StringUtils.isEmpty(keystore_password) && !keyPassword.equals(keystore_password))
+        {
+            throw new IllegalArgumentException("'keystore_password' and 'key_password' both configurations are given and the " +
+                                   "values don't match");
```
don't -> do not, it's on the next warn -> it is

I think it is better to stick to the "proper" English. It does not kill anybody to read "it is" instead of "it's". There is nowhere to hurry :)

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+        {
+            keyPassword = keystore_password;
+        }
+        else if (!StringUtils.isEmpty(keystore_password) && !keyPassword.equals(keystore_password))
```
why do we actually need two variables with seemingly same names and there we are doing a logic around their possible equality? That is quite confusing to go through if you ask me. Or "keystore" and "key" passwords are password for completely different things? I would welcome if we put more Javadoc on top of relevant properties / methods so we can know without any ambiguity which is which and for what (it they are meant to stay).

maulin-vasavada replied:
I did think about it when I wrote it but the challenge is - with PEM we have Private Key with password in the input vs with JKS we have `keystore`. So I had a choice to stick to existing configurations for `keystore_password` but I thought of adding `private_key_password` configuration for the `PEMBasedSslContextFactory` so users can relate to the PEM world and ignore keystore_password altogether. However since we already have keystore_password, I thought leaving the option to use that in case might be helpful if people just convert the keys to the PEM format and keep the password same in the keystore_password. 

I am open to dropping the support for reading keystore_password configuration for the PEM based keys and strictly just require `keyPassword` variable name and `private_key_password` configuration parameter. Also, the variable name with underscore came from the FileBasedSslContextFactory and I am open to change the variable names in that class to remove underscore and follow the camel case.

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+            {
+                throw new SSLException("Can't build KeyManagerFactory in absence of the Private Key");
+            }
+        } catch(Exception e) {
```
space

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+            {
+                throw new SSLException("Can't build KeyManagerFactory in absence of the trusted certificates");
+            }
+        } catch(Exception e) {
```
same

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+
+                TrustManagerFactory tmf = TrustManagerFactory.getInstance(
+                algorithm == null ? TrustManagerFactory.getDefaultAlgorithm() : algorithm);
+                KeyStore ts = buildTrustStore();
```
maybe this makes perfect sense from the implementation perspective, I am just saying that if I am _building a truststore_, I do not expect that I assign it to a variable which has type `KeyStore`.

maulin-vasavada replied:
Yes, I agree with you but since Java only offers type `KeyStore` for both -keystore and truststore the code is like this.

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+            }
+            else
+            {
+                throw new SSLException("Can't build KeyManagerFactory in absence of the trusted certificates");
```
Can not

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+                throw new SSLException("Can't build KeyManagerFactory in absence of the trusted certificates");
+            }
+        } catch(Exception e) {
+            throw new SSLException("failed to build trust manager store for secure connections", e);
```
some messages starts with lower case, some with upper case, cant be this same one style or the other everywhere?

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+    private String readPEMFile(String file) throws IOException
+    {
+        StringBuilder fileData = new StringBuilder();
+        try(BufferedReader br = new BufferedReader(new FileReader(file)))
```
space

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+        try(BufferedReader br = new BufferedReader(new FileReader(file)))
+        {
+            String line = null;
+            while( (line = br.readLine()) != null)
```
`while ((line = br.readLine()) != null)`

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+        }
+    }
+
+    private String readPEMFile(String file) throws IOException
```
cant be the logic in this method just replaced by this?

```
return new String(Files.readAllBytes(Paths.get(file)));
```

I would also catch the exception and logged it what happened and then rethrow if necessary.

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+
+        KeyStore keyStore = KeyStore.getInstance(targetStoreType);
+        keyStore.load(null, null);
+        keyStore.setKeyEntry("cassandra-ssl-keystore", privateKey, keyPasswordArray, certChainArray);
```
is "cassandra-ssl-keystore" going to be a constant here?

maulin-vasavada replied:
yes. So what is the suggestion? To move this to a constant?

smiklosovic commented on src/java/org/apache/cassandra/security/PEMBasedSslContextFactory.java:
```diff
@@ -0,0 +1,343 @@
+
+        KeyStore keyStore = KeyStore.getInstance(targetStoreType);
+        keyStore.load(null, null);
+        for (int i=0; i < certChainArray.length; i++) {
```
mostly spaces and brackets

```
for (int i = 0; i < certChainArray.lenght; i++)
{
    keyStore.setCertificateEntry("blabla" + (i + 1), ...);
}
```

smiklosovic commented on test/conf/cassandra-pem-jks-sslcontextfactory.yaml:
```diff
@@ -0,0 +1,151 @@
+        class_name: org.apache.cassandra.security.PEMBasedSslContextFactory
+        parameters:
+          private_key: |
+            -----BEGIN ENCRYPTED PRIVATE KEY-----
```
what is in these keys? what is their content, how they look like in plain text? some dump would help, for debugging purposes.

maulin-vasavada replied:
The content of this is PKCS#8 formatted Private Key (either RSA, DSA or EC based).

In the PEMReader class in this PR there are debug statements to print potentially useful information about the Private Key. PEMReader also prints the useful details for Trusted certs (based my experience dealing with SSL debugging issues). However with Private Key I was more conservative.

Please let me know if this makes sense or we should do something specific here.

driftx committed:
Use correct clock source in TimestampSerializerTest

Patch by brandonwilliams; reviewed by edimitrova for CASSANDRA-17120

driftx committed:
Replace System.currentTimeMillis with Clock.Global.currentTimeMillis

Patch by brandonwilliams; reviewed by bereng for CASSANDRA-17123

dcapwell committed:
JVMStabilityInspector.forceHeapSpaceOomMaybe should handle all non-heap OOMs rather than only supporting direct only

patch by David Capwell; reviewed by Caleb Rackliffe, Yifan Cai for CASSANDRA-17128

dcapwell committed:
MessagingServiceTest listenOptionalSecureConnection and listenRequiredSecureConnection fail sporadically

patch by David Capwell; reviewed by Benedict Elliott Smith, Caleb Rackliffe for CASSANDRA-17033

Commit:
Addressing formatting issues

Commit:
Merge branch 'CASSANDRA-17031' of https://github.com/maulin-vasavada/cassandra into CASSANDRA-17031

Commit:
Moving the inline PEM data in test to test resources as files

maulin-vasavada commented:
Closing this PR due to messed up commit history. Raised a new PR - https://github.com/apache/cassandra/pull/1316
